### PR TITLE
ENH: optimize: add DF-SANE nonlinear derivative-free solver

### PR DESCRIPTION
**Pull request by pv**
Add DF-SANE, a derivative-free nonlinear equation system solving method.

Currently, the other large-scale derivative-free method offered in Scipy are Newton-Krylov and the Broyden/Anderson family. As one data point, solving a certain fixed-point integrodifferential equation problem, DF-SANE performed significantly better than the above solvers (NK was very slow as the Jacobian was 'difficult', Broyden/Anderson had problems in late-stage convergence).

In R, dfsane is implemented in the BB package, not sure how many other numerical computing environments include it.

Comments welcome.

### CODE REVIEW

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -3142,6 +3142,36 @@ def show_options(solver=None, method=None):
...
+        fnorm : callable, optional
+            Norm to use in the convergence check. If None, 2-norm is used.
+
+        maxfev : int, optional
+            Maximum number of function evaluations.
+
+        disp : bool, optional
+            Whether to print convergence process to stdout.
+
+        eta_stragegy : callable, optional
```
Comment (argriffing) [inline scipy/optimize/optimize.py]: spelling

File: scipy/optimize/_spectral.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        # vs. scaling of F.
+        #
+        # This choice satisfies the requirements in the paper.  f_k is
+        # bounded: the maximum allowed by the line search is
+        #
+        #     f_{k+1} = f_k * [1 + 1 / (1 + k)**2]
+        #
+        # Now, prod(1 + 1/(1 + k)**2, k=0..inf) = sinh(pi)/(2*pi) < inf.
+        def eta_strategy(k, x, F):
+            return f_k / (1 + k)**2
```
Comment (dlax) [inline scipy/optimize/_spectral.py]: `f_k` ?

Comment (pv) [inline scipy/optimize/_spectral.py]: Dirty trick -- it comes from outer scope, avoids computation of 2-norm.

Comment (charris) [inline scipy/optimize/_spectral.py]: Maybe add a comment?

**Comment by argriffing** [comment]
TravisCI failures are a few `E261 at least two spaces before inline comment`, and a solver accuracy test fails on old numpy.

**Comment by pv** [comment]
Fixing things up. Didn't manage to reproduce the failure with old numpy + 2.6, so it's probably due to variation in rounding errors --- the nfev limit is exceeded only by one.

**Comment by pv** [comment]
No idea what's going on with that one failing test --- trying to abuse Travis to see what's going on...

**Comment by argriffing** [comment]
With numpy 1.6.2 and scipy 0.16.0.dev+a4c6762 the `-s optimize` tests pass on my system (linux 64 bits).

```
iter 0: ||F|| = 48.0093, sigma = 1
iter 1: ||F|| = 59.4557, sigma = -139.41
iter 2: ||F|| = 42.7095, sigma = 1.54805
iter 3: ||F|| = 46.7212, sigma = -1.76338
iter 4: ||F|| = 40.378, sigma = -1.19454
iter 5: ||F|| = 33.9095, sigma = -0.494159
iter 6: ||F|| = 43.4696, sigma = 9.90173
iter 7: ||F|| = 40.704, sigma = 1.51374
iter 8: ||F|| = 34.9705, sigma = 1.00557
iter 9: ||F|| = 31.6877, sigma = 1.02115
iter 10: ||F|| = 41.1833, sigma = -1.85206
...
iter 160: ||F|| = 1.12335, sigma = -0.3087
iter 161: ||F|| = 0.417701, sigma = -0.454411
iter 162: ||F|| = 0.64129, sigma = 0.849987
iter 163: ||F|| = 0.0825631, sigma = 0.754565
iter 164: ||F|| = 0.0772665, sigma = 3.88443
iter 165: ||F|| = 0.993951, sigma = -0.327476
iter 166: ||F|| = 0.0318479, sigma = -0.317522
iter 167: ||F|| = 0.0135234, sigma = -0.501424
iter 168: ||F|| = 0.0223474, sigma = 0.768472
iter 169: ||F|| = 0.000223768, sigma = 0.77603
```

**Comment by pv** [comment]
The solution is not numerically stable for the default `eta_strategy`, perturbing the initial guess makes convergence fail also here. It seems it's up to rounding errors when it starts to converge.

A slower `eta_strategy` with exponent 1.2 instead of 2 converges apparently better for this case (any exponent > 1 satisfies the constraints). What to choose here doesn't seem to be well understood, apart from the boundedness constraint; the paper [1] just gives one approach without comment and the Fortran code also has commented-out schedules... IMHO a default value that is invariant vs scaling of the objective function is desirable, so I'd rather not use those guesses.

**Comment by argriffing** [comment]
Regarding style there are a few loose `assert` which don't use `numpy.testing`, and some of the equality or inequality `assert_` could be more specific (`assert_equal`, `assert_array_less`).  The new nonlinear solver itself looks good.

**Comment by pv** [comment]
Fixed up some style issues. Using `assert_array_less` leads to ugliness esp. in the floating point case since I'd like to test for `<=` and not `<` :(  so I didn't replace those.

Also, changed my mind again on the default `eta_strategy`. Basing the schedule on the initial residual seems to be much better (for the F4 function and also for my real-world problem --- convergence vs. no convergence), apparently because in late convergence stage it's better to just rely on the BB spectral step length rather than waste time on line searching. The current choice is scale invariant variant of the choice in the DF-SANE paper, and is the same for functions with norms of order 1.

File: scipy/optimize/_spectral.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    ----------
+    .. [1] "Spectral residual method without gradient information for solving
+           large-scale nonlinear systems of equations." W. La Cruz,
+           J.M. Martinez, M. Raydan. Math. Comp. **75**, 1429 (2006).
+    .. [2] W. La Cruz, Opt. Meth. Software, 29, 24 (2014).
+
+    """
+    _check_unknown_options(unknown_options)
+    if jac is not None:
+        warnings.warn("DF-SANE solver does not use a Jacobian", category=OptimizeWarning)
```
Comment (dlax) [inline scipy/optimize/_spectral.py]: Wouldn't it be better to check this in `root` and not pass `jac` as this is done for other solvers?

**Comment by argriffing** [comment]
> Using `assert_array_less` leads to ugliness esp. in the floating point case since I'd like to test for `<=` and not `<` :( so I didn't replace those.

This is OK with me.  I've been annoyed with that difference in the past as well.

File: scipy/optimize/_spectral.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            v = np.asarray(func(z, *args)).ravel()
+            return _complex2real(v)
+
+        x0 = _complex2real(x0)
+        F = _complex2real(F)
+    else:
+        def wrap_func(x, *a, **kw):
+            x = x.reshape(x0_shape)
+            return np.asarray(func(x, *args)).ravel()
+
```
Comment (dlax) [inline scipy/optimize/_spectral.py]: Depending on `is_complex`, `wrap_func` has a different signature.

File: scipy/optimize/_spectral.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    func, x0, x0_shape, F_k, is_complex = _wrap_func(func, x0, args)
+
+    def f(x):
+        if nfev[0] >= maxfev:
+            raise _NoConvergence()
+
+        F = func(x)
+        f = np.linalg.norm(F)**nexp
+        nfev[0] += 1
+        return f, F
```
Comment (dlax) [inline scipy/optimize/_spectral.py]: Could the two wrappings `_wrap_func` and `f` here be combined into a single one? I'll make reading easier IMHO.

**Comment by pv** [comment]
Cleaned up the code a bit to address comments.
.
Also added an alternative line search from [3]. Some articles citing [3]
claimed it may work better in practice, so I'll just include it too, but
not as the default. Things here are a bit sensitive still on how the
eta_strategy is chosen.

**Comment by dlax** [comment]
Merged and mention this to 0.16 release notes, thanks @pv !

**Comment by pv** [comment]
That the tests showed that convergence is sensitive to rounding errors, and that there are some difficulties in reproducing exact iteration counts, is apparently not a coincidence: http://link.springer.com/article/10.1007/s10915-011-9521-3 --- I guess it just inherits the chaotic properties of the BB/LSD method.

**Pull request merged**